src/mscorlib/shared/System/Span.cs

```diff
@@ -408,5 +408,47 @@ public T[] ToArray()
         /// Returns an empty <see cref="Span{T}"/>
         /// </summary>
         public static Span<T> Empty => default(Span<T>);
+
+        /// <summary>Gets an enumerator for this span.</summary>
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        /// <summary>Enumerates the elements of a <see cref="Span{T}"/>.</summary>
+        public ref struct Enumerator
+        {
+            /// <summary>The span being enumerated.</summary>
+            private readonly Span<T> _span;
+            /// <summary>The next index to yield.</summary>
+            private int _index;
+
+            /// <summary>Initialize the enumerator.</summary>
+            /// <param name="span">The span to enumerate.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal Enumerator(Span<T> span)
+            {
+                _span = span;
+                _index = -1;
+            }
+
+            /// <summary>Advances the enumerator to the next element of the span.</summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool MoveNext()
+            {
+                int index = _index + 1;
+                if (index < _span.Length)
+                {
+                    _index = index;
+                    return true;
+                }
+
+                return false;
+            }
+
+            /// <summary>Gets the element at the current position of the enumerator.</summary>
+            public ref T Current
```

Comment:
>public ref T Current [](start = 12, length = 20)

Interestingly this works!! And it does comply with the spec of the foreach pattern - In a way that the spec does not demand that Current returns by value (I think it does not).
I think this is a good thing, but it seems a bit accidental that this works :-) . We should at least add tests on the compiler side.  @MadsTorgersen 

And should think about byref foreach too. Seems like a good thing to allow.

Reply:
> Interestingly this works!!

Yup:
https://github.com/dotnet/corefx/pull/25124/files#diff-6f98a2e0a1bc4eca989a22cb3f170e13R27
https://github.com/dotnet/corefx/pull/25124/files#diff-59c8f4cdb3ffcc4e7e82ce4a3db9f414R27